In [1]:
import pandas as pd
import numpy as np

In [2]:
file1  = pd.read_excel(r'C:\Users\ASUS\Documents\Class\Applied Forecasting Model\Kansas City Chiefs_clean.xlsx')
file2  = pd.read_excel(r'C:\Users\ASUS\Documents\Class\Applied Forecasting Model\Philadelphia Eagles_clean.xlsx')

In [3]:
#use average of playoffs as predict data
test1 = file1.iloc[[17,18,36,37,38,55,56,57],:].groupby("Team").mean().reset_index(drop=False)
test1

Team   Score  1stD_Offense  TotYd_Offense  PassY_Offense  \
0  Kansas City Chiefs  28.375        25.125        418.875         299.25   

   RushY_Offense  TO_Offense  1stD_Defense  TotYd_Defense  PassY_Defense  \
0        119.625       1.125          21.5        338.375        228.125   

   RushY_Defense  TO_Defense  
0         110.25       1.125

In [4]:
#use average of playoffs as predict data
test2 = file2.iloc[[17,18,36],:].groupby("Team").mean().reset_index(drop=False)
test2

Team  Score  1stD_Offense  TotYd_Offense  PassY_Offense  \
0  Philadelphia Eagles   28.0     21.666667     341.333333          171.0   

   RushY_Offense  TO_Offense  1stD_Defense  TotYd_Defense  PassY_Defense  \
0     170.333333         1.0     15.666667     246.666667          145.0   

   RushY_Defense  TO_Defense  
0     101.666667    1.333333

In [5]:
file1.head()

Team  Score  1stD_Offense  TotYd_Offense  PassY_Offense  \
0  Kansas City Chiefs     44            33            488            360   
1  Kansas City Chiefs     27            15            319            226   
2  Kansas City Chiefs     17            20            315            257   
3  Kansas City Chiefs     41            27            417            228   
4  Kansas City Chiefs     30            29            368            265   

   RushY_Offense  TO_Offense  1stD_Defense  TotYd_Defense  PassY_Defense  \
0            128           1            18            282            179   
1             93           0            21            401            326   
2             58           2            19            259            177   
3            189           1            27            376            373   
4            103           0            18            378            223   

   RushY_Defense  TO_Defense  
0            103           0  
1             75           1  
2             82           1  
3              3           2  
4            155           0

In [6]:
file2.head()

Team  Score  1stD_Offense  TotYd_Offense  PassY_Offense  \
0  Philadelphia Eagles     38            27            455            239   
1  Philadelphia Eagles     24            25            486            323   
2  Philadelphia Eagles     24            21            400            328   
3  Philadelphia Eagles     29            25            401            191   
4  Philadelphia Eagles     20            24            357            218   

   RushY_Offense  TO_Offense  1stD_Defense  TotYd_Defense  PassY_Defense  \
0            216           0            23            386            205   
1            163           1            20            264            202   
2             72           0            20            240            153   
3            210           1            13            219            148   
4            139           0            23            363            239   

   RushY_Defense  TO_Defense  
0            181           1  
1             62           3  
2             87           1  
3             71           5  
4            124           1

## Linear regression

In [7]:
#team1 score: Kansas
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

X_linear = file1.iloc[:,2:]
y_linear = file1.iloc[:,1]

#split data into training and testing
X_linear_train, X_linear_test, y_linear_train, y_linear_test = train_test_split(X_linear, y_linear, test_size=0.2, random_state=5)

reg1 = LinearRegression().fit(X_linear_train, y_linear_train)
y_pred1 = reg1.predict(X_linear_test)

def mape(y_actual, y_predicted):
    mape = np.mean(np.abs((y_actual - y_predicted)/y_actual))
    return mape

print('prediction for Kansas: ', round(reg1.predict(test1.iloc[:,2:])[0],0))
print('accuracy rate: ', round(reg1.score(X_linear_test, y_linear_test),2))
print('MAPE: ', round(mape(y_linear_test, y_pred1),2))
print('MSE: ' , round(mean_squared_error(y_linear_test, y_pred1),2))

prediction for Kansas:  29.0
accuracy rate:  0.65
MAPE:  0.14
MSE:  23.69


In [8]:
#team2 score: Philadelphia
X_linear2 = file2.iloc[:,2:]
y_linear2 = file2.iloc[:,1]

#split data into training and testing
X_linear_train2, X_linear_test2, y_linear_train2, y_linear_test2 = train_test_split(X_linear2, y_linear2, test_size=0.2, random_state=55)

reg2 = LinearRegression().fit(X_linear_train2, y_linear_train2)
y_pred2 = reg1.predict(X_linear_test2)

print('prediction for Philadelphia: ', round(reg2.predict(test2.iloc[:,2:])[0],0))
print('accuracy rate: ', round(reg2.score(X_linear_test2 , y_linear_test2),2))
print('MAPE: ', round(mape(y_linear_test2, y_pred2),2))
print('MSE: ' , round(mean_squared_error(y_linear_test2, y_pred2),2))

prediction for Philadelphia:  30.0
accuracy rate:  0.62
MAPE:  0.24
MSE:  33.73


## Neural network

In [9]:
#team1 score: Kansas
from sklearn.neural_network import MLPClassifier

X_neu1 = file1.iloc[:,2:]
y_neu1 = file1.iloc[:,1]

#split data into training and testing
X_neu_train, X_neu_test, y_neu_train, y_neu_test = train_test_split(X_neu1, y_neu1, test_size=0.2, random_state=10)

clf_new1 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(10, 2), random_state=1)
clf_new1.fit(X_neu_train, y_neu_train)
MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1, solver='lbfgs')
y_pred1 = clf_new1.predict(X_neu_test)

print('prediction for Kansas: ', round(clf_new1.predict(test1.iloc[:,2:])[0],0))
print('accuracy rate: ', round(clf_new1.score(X_neu_test, y_neu_test),2))
print('MAPE: ', round(mape(y_pred1, y_neu_test),2))
print('MSE: ' , round(mean_squared_error(y_pred1, y_neu_test),2))

prediction for Kansas:  27
accuracy rate:  0.08
MAPE:  0.34
MSE:  127.58


In [10]:
#team2 score: Philadelphia
from sklearn.neural_network import MLPClassifier

X_neu2 = file2.iloc[:,2:]
y_neu2 = file2.iloc[:,1]

#split data into training and testing
X_neu_train2, X_neu_test2, y_neu_train2, y_neu_test2 = train_test_split(X_neu2, y_neu2, test_size=0.2, random_state=55)

clf_new2 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(10, 2), random_state=1)
clf_new2.fit(X_neu_train2, y_neu_train2)
MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1, solver='lbfgs')
y_pred2 = clf_new2.predict(X_neu_test2)

print('prediction for Philadelphia: ', round(clf_new2.predict(test2.iloc[:,2:])[0],0))
print('accuracy rate: ', round(clf_new2.score(X_neu_test2, y_neu_test2),2))
print('MAPE: ', round(mape(y_pred2, y_neu_test2),2))
print('MSE: ' , round(mean_squared_error(y_pred2, y_neu_test2),2))

prediction for Philadelphia:  17
accuracy rate:  0.09
MAPE:  0.49
MSE:  96.55


## Random forest

In [11]:
#team 1 score : Kansas
from sklearn.ensemble import RandomForestRegressor

X_rf1 = file1.iloc[:,2:]
y_rf1 = file1.iloc[:,1]

#split data into training and testing
X_rf_train, X_rf_test, y_rf_train, y_rf_test = train_test_split(X_rf1, y_rf1, test_size=0.3, random_state=20)

rf1 = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf1.fit(X_rf_train, y_rf_train)
y_pred1 = rf1.predict(X_rf_test)

print("prediction of Kansas: ", round(rf1.predict(test1.iloc[:,2:])[0],0))
print('accuracy rate: ', round(rf1.score(X_rf_test, y_rf_test),0))
print('MAPE: ', round(mape(y_rf_test, y_pred1),2))
print('MSE: ' , round(mean_squared_error(y_rf_test, y_pred1),2))

prediction of Kansas:  28.0
accuracy rate:  0.0
MAPE:  0.2
MSE:  35.95


In [12]:
#team 2 score : Philadelphia
X_rf2 = file2.iloc[:,2:]
y_rf2 = file2.iloc[:,1]

#split data into training and testing
X_rf_train2, X_rf_test2, y_rf_train2, y_rf_test2 = train_test_split(X_rf2, y_rf2, test_size=0.25, random_state=30)

rf2 = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf2.fit(X_rf_train2, y_rf_train2)
y_pred2 = rf1.predict(X_rf_test2)

print("prediction of Philadelphia: ", round(rf2.predict(test2.iloc[:,2:])[0],0))
print('accuracy rate: ', round(rf2.score(X_rf_test2, y_rf_test2),0))
print('MAPE: ', round(mape(y_rf_test2, y_pred2),2))
print('MSE: ' , round(mean_squared_error(y_rf_test2, y_pred2),2))

prediction of Philadelphia:  25.0
accuracy rate:  0.0
MAPE:  0.2
MSE:  31.71


## KNN

In [13]:
#team 1 score : Kansas
from sklearn.neighbors import KNeighborsRegressor

X_knn1 = file1.iloc[:,2:]
y_knn1 = file1.iloc[:,1]

#split data into training and testing
X_knn_train, X_knn_test, y_knn_train, y_knn_test = train_test_split(X_knn1, y_knn1, test_size=0.15, random_state=60)

knn_model1 = KNeighborsRegressor(n_neighbors=3)
knn_model1.fit(X_knn_train, y_knn_train)
y_pred = knn_model1.predict(X_knn_test)

print("prediction of Kansas: ", round(knn_model1.predict(test1.iloc[:,2:])[0],0))
print('accuracy rate: ', round(knn_model1.score(X_knn_test, y_knn_test),0))
print('MAPE: ', round(mape(y_knn_test, y_pred),2))
print('MSE: ' , round(mean_squared_error(y_knn_test, y_pred),2))

prediction of Kansas:  26.0
accuracy rate:  1.0
MAPE:  0.7
MSE:  57.72


In [14]:
#team 2 score : Philadelphia
X_knn2 = file2.iloc[:,2:]
y_knn2 = file2.iloc[:,1]

#split data into training and testing
X_knn_train2, X_knn_test2, y_knn_train2, y_knn_test2 = train_test_split(X_knn2, y_knn2, test_size=0.15, random_state=30)

knn_model2 = KNeighborsRegressor(n_neighbors=3)
knn_model2.fit(X_knn_train2, y_knn_train2)
y_pred2 = knn_model2.predict(X_knn_test2)

print("prediction of Philadelphia: ", round(knn_model2.predict(test2.iloc[:,2:])[0],0))
print('accuracy rate: ', round(knn_model2.score(X_knn_test2, y_knn_test2),0))
print('MAPE: ', round(mape(y_knn_test2, y_pred2),2))
print('MSE: ' , round(mean_squared_error(y_knn_test2, y_pred2),2))

prediction of Philadelphia:  17.0
accuracy rate:  0.0
MAPE:  0.16
MSE:  22.39


## Naive Bayes

In [15]:
#team 1 score : Kansas
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, accuracy_score

X_naive1 = file1.iloc[:,2:]
y_naive1 = file1.iloc[:,1]

#split data into training and testing
X_nb_train, X_nb_test, y_nb_train, y_nb_test = train_test_split(X_naive1, y_naive1, test_size=0.15, random_state=60)

naive_classifier1 = GaussianNB()
naive_classifier1.fit(X_nb_train, y_nb_train)
y_pred1 = naive_classifier1.predict(X_nb_test)

print("prediction of Kansas: ", round(naive_classifier1.predict(test1.iloc[:,2:])[0],0))
print('accuracy rate: ', round(ac1,2))
print('MAPE: ', round(mape(y_nb_test, y_pred1),2))
print('MSE: ' , round(mean_squared_error(y_nb_test, y_pred1),2))

prediction of Kansas:  27
accuracy rate:  0.11
MAPE:  0.95
MSE:  99.56


In [16]:
#team 2 score : Philadelphia
X_naive2 = file2.iloc[:,2:]
y_naive2 = file2.iloc[:,1]

#split data into training and testing
X_nb_train2, X_nb_test2, y_nb_train2, y_nb_test2 = train_test_split(X_naive2, y_naive2, test_size=0.1, random_state=60)

naive_classifier2 = GaussianNB()
naive_classifier2.fit(X_nb_train2, y_nb_train2)
y_pred2 = naive_classifier2.predict(X_nb_test2)

print("prediction of Philadelphia: ", round(naive_classifier2.predict(test2.iloc[:,2:])[0],0))
print('accuracy rate: ', round(ac2,2))
print('MAPE: ', round(mape(y_nb_test2, y_pred2),2))
print('MSE: ' , round(mean_squared_error(y_nb_test2, y_pred2),2))

prediction of Philadelphia:  29
accuracy rate:  0.17
MAPE:  0.26
MSE:  37.67


# Conclusion


**Linear regression: 
<br>
<br>
Kansas accuracy: 0.65
<br>
Philadelphia accuracy: 0.62
<br>
score(Kansas v.s. Philadelphia): 29:30**
***


**Neural network: 
<br>
<br>
Kansas accuracy: 0.08
<br>
Philadelphia accuracy: 0.09
<br>
score(Kansas v.s. Philadelphia): 27:17**
***

**Random forest: 
<br>
<br>
Kansas accuracy: 0.38
<br>
Philadelphia accuracy: 0.49
<br>
score(Kansas v.s. Philadelphia): 28:25**
***

**KNN: 
<br>
<br>
Kansas accuracy: 0.53
<br>
Philadelphia accuracy: 0.41
<br>
score(Kansas v.s. Philadelphia): 26:17**
***

**Naive Bayes: 
<br>
<br>
Kansas accuracy: 0.11
<br>
Philadelphia accuracy: 0.17
<br>
score(Kansas v.s. Philadelphia): 27:29**
***

## Personal Prediction:

## score(Kansas v.s. Philadelphia): 26:28